### PR TITLE
chore: Modified getToken flow for userpool signIn

### PR DESCRIPTION
**Pull request by royjit**
*Issue #, if available:*

*Description of changes:*

PR converts the re-write the `getTokens` flow such that it is decoupled from the internal AWSMobileClient logic. The previous implementation was tightly coupled with the signIn flow and uses many common properties. 

With this PR the `getTokens` call for `userPools` is moved to a separate operation named `FetchUserPoolTokensOperation`. This operation uses a separate instance of user pool client instead of the single client used throughout in AWSMobileClient. Another change from the previous implementation is that the user pool client does not wait for the user to signIn if refresh token expiry, instead the operation waits. And when the user signIn again, the operation will try to fetch the token again. 

Rough sequence diagram of a getTokens logic:

```mermaid
sequenceDiagram
    autonumber
    participant User
    participant AWSMC as AWSMobileClient
    participant Op as FetchUserPoolTokensOperation
    participant SDK as CognitoSDK
    User->>AWSMC: getTokens
    Note over AWSMC,Op: create Operation and add to a serial queue
    AWSMC->>+Op: 
    Op->>Op: startOperation
    Op->>SDK: Invoke getSession
    alt getSession return token
          SDK-->>Op: Token
          Op-->>User:Invoke completion with token
     else getSession return error
          SDK-->>Op: Error
           alt error is delegate missing
                 Op-->>AWSMC: Invoke delegate tokenFetchNeedsAuthentication
                 alt user signIn
                       User->>AWSMC: signIn
                       AWSMC->>Op: Inform signedIn
                       Op->>Op: Got to step 4
                  else user signedOut
                        User->>AWSMC: signOut
                       AWSMC->>Op: Inform signedOut
                       Op-->>User:Invoke completion with error
                  else user releaseSignInWait
                        User->>AWSMC: releaseSignInWait
                       AWSMC->>Op: Inform releaseWait
                       Op-->>User:Invoke completion with error
                  end
           else some other error
                 Op-->>User:Invoke completion with error
           end
     end
```

Internally the operation handles the logic through a state machine:

```mermaid
  flowchart TD;
     notStarted((notStarted))-->|startOperation| fetching;
     notStarted-->|releaseWait| releasedSignInWait;
     releasedSignInWait-->|startOperation|error((error));
     fetching-->|tokenFetched|fetched((fetched)) ;
     fetching-->|serviceError|error ;
     fetching-->|signedOut|error ;
     fetching-->|noUserFound|error ;
     fetching-->|tokenExpired|waitingForSignIn ;
     waitingForSignIn-->|signedIn|fetching ;
     waitingForSignIn-->|releaseWait|error ;
     waitingForSignIn-->|signedOut|error ;
```

*Check points:*

- [x] Added new tests to cover change, if needed
- [x] All unit tests pass
- [x] All integration tests pass
- [ ] Updated CHANGELOG.md
- [ ] Documentation update for the change if required
- [x] PR title conforms to conventional commit style

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by atierian — DISMISSED** [review]
If this is the decoupling change, can we add a few sentences in the PR with a high level explanation of the changes as a future reference?

Otherwise LGTM.

*Post merge we should double check that the copyright notice (specifically `Copyright 2017-2018`) is correct.

File: AWSAuthSDK/Sources/AWSMobileClient/Models/Tokens.swift [hunk 1]
```diff
@@ -31,3 +32,34 @@ public struct Tokens {
...
+        if let token = self.accessToken {
+            accessToken = SessionToken(tokenString: token.tokenString)
+        }
+        if let token = self.refreshToken {
+            refreshToken = SessionToken(tokenString: token.tokenString)
+        }
+        return Tokens(idToken: idToken,
+                      accessToken: accessToken,
+                      refreshToken: refreshToken,
+                      expiration: self.expirationTime)
```
Comment (harsh62) [inline AWSAuthSDK/Sources/AWSMobileClient/Models/Tokens.swift:54]: Could be just 
```
        return Tokens(idToken: SessionToken(tokenString: self.idToken?.tokenString),
                      accessToken: SessionToken(tokenString: self.accessToken?.tokenString),
                      refreshToken: SessionToken(tokenString: self.refreshToken?.tokenString),
                      expiration: self.expirationTime)
```

**Review by harsh62 — DISMISSED** [review]
LGTM, with minor comments..

File: AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientOperations/AWSAsyncOperations/AWSAsyncOperation.swift [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        state == .cancelled
+    }
+
+    override var isAsynchronous: Bool { true }
+
+    override func start() {
+        guard !isCancelled else {
+            transition(to: .finished)
+            return
+        }
```
[the remark this answers is not in the document]
Comment (royjit) [reply AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientOperations/AWSAsyncOperations/AWSAsyncOperation.swift:61]: that is expected, it means that the cancelled the operation. You are not expected to get a result from a cancelled operation.

File: AWSAuthSDK/Sources/AWSMobileClient/Models/Tokens.swift (hunk 1 above)
Comment (royjit) [reply AWSAuthSDK/Sources/AWSMobileClient/Models/Tokens.swift]: Changed

**Review by harsh62 — APPROVED** [review]

**Pull request merged**